torch/lowrank.py

```diff
@@ -0,0 +1,249 @@
+
+    return Q
+
+
+def svd(A, q=6, niter=2, M=None):
+    """Return the singular value decomposition ``(U, S, V)`` of a matrix,
+    batches of matrices, or a sparse matrix :math:`A` such that
+    :math:`A \approx U diag(S) V^T`. In case :math:`M` is given, then
+    SVD is computed for the matrix :math:`A - M`.
+
+    .. note:: The implementation is based on the Algorithm 5.1 from
+              Halko et al, 2009.
+
+    .. note:: To obtain repeatable results, reset the seed for the
+              pseudorandom number generator
+
+    .. note:: The input is assumed to be a low-rank matrix.
+
+    .. note:: In general, use the full-rank SVD implementation
+              ``torch.svd`` for dense matrices due to its 10-fold
+              higher performance characteristics. The low-rank SVD
+              will be useful for huge sparse matrices that
+              ``torch.svd`` cannot handle.
+
+    Arguments::
+        A (Tensor): the input tensor of size :math:`(*, m, n)`
+
+        q (int, optional): a slightly overestimated rank of A.
+
+        niter (int, optional): the number of subspace iterations to
+                               conduct; niter must be a nonnegative
+                               integer, and defaults to 2
+
+        M (Tensor, optional): the input tensor's mean of size
+                              :math:`(*, 1, n)`.
+
+    References::
+        - Nathan Halko, Per-Gunnar Martinsson, and Joel Tropp, Finding
+          structure with randomness: probabilistic algorithms for
+          constructing approximate matrix decompositions,
+          arXiv:0909.4061 [math.NA; math.PR], 2009 (available at
+          `arXiv <http://arxiv.org/abs/0909.4061>`_).
+
+    """
+    m, n = A.shape[-2:]
+    matmul = _utils.get_matmul(A)
+    if M is None:
+        M_t = None
+    else:
+        matmul2 = _utils.get_matmul(M)
+        M_t = _utils.transpose(M)
+    A_t = _utils.transpose(A)
+
+    # Algorithm 5.1 in Halko et al 2009, slightly modified to reduce
+    # the number conjugate and transpose operations
+    if m < n:
+        # computing the SVD approximation of a transpose in order to
+        # keep B shape minimal
+        Q = get_approximate_basis(A_t, q, niter=niter, M=M_t)
+        Q_c = _utils.conjugate(Q)
+        if M is None:
+            B_t = matmul(A, Q_c)
+        else:
+            B_t = matmul(A, Q_c) - matmul2(M, Q_c)
+        U, S, V = torch.svd(B_t)
+        V = Q.matmul(V)
+    else:
+        Q = get_approximate_basis(A, q, niter=niter, M=M)
+        Q_c = _utils.conjugate(Q)
+        if M is None:
+            B = matmul(A_t, Q_c)
+        else:
+            B = matmul(A_t, Q_c) - matmul2(M_t, Q_c)
+        U, S, V = torch.svd(_utils.transpose(B))
+        U = Q.matmul(U)
+
+    return U, S, V
+
+
+def pca(A, q=None, center=True, niter=2):
+    r"""Performs Principal Component Analysis (PCA) on a low-rank matrix,
+    batches of such matrices, or sparse matrix.
```

Comment:
Maybe it is worth mentioning that the PCA in here is a linear PCA.

Reply:
Changed to `Performs linear Principal Component Analysis...`.

Note that the `pca` function here does not attempt to provide a full-featured PCA tool but is rather an example application of low-rank SVD. I think to achieve the equivalent PCA functionality that, say, sklearn provides would be a much bigger project than this PR can be.

Reply:
Ok, got it. The description says that the samples are sparse, but then the covariance matrix is not. This means that one will likely not be able to run the algorithm for his small to medium, say, sparse genomic data. If, however, the implementation operates on the kernel matrix, then it should be feasible. All the ingredients for my own version of specific, utilizing structure in my data, PCA are already here, namely SVD. Then, why would I use this specific implementation?